File: slack/process-picks.pl
#!/usr/bin/env perl
use strict;
use warnings;

use v5.10;

our $VERSION = 0.1;

use utf8;
use Spreadsheet::XLSX;
use Text::Unidecode;
use List::Util qw(min max sum);
use Data::Printer;
use Getopt::Long;

sub makeSlug {
  my ($art, $alb) = @_;

  $art = unidecode(lc($art));
  $art =~ s/[^a-z0-9]+/_/gxsm;
  $alb = unidecode(lc($alb));
  $alb =~ s/[^a-z0-9]+/_/gxsm;
  return "$art-$alb";
}

sub showUsage {
  my ($msg) = @_;

  print STDERR <<__USAGE;
EOY List Picks Processor

./process-picks.pl --source <source.xlsx>
__USAGE

  if ($msg) {
    say STDERR $msg;
  }
}

my $source;
my $skip = 0; # rows to skip
GetOptions (
  "source=s" => \$source,
  "skip=i" => \$skip
) or die("Error in command line arguments\n");

if (!$source || ! -f $source) {
  showUsage('Cannot find source or none specified');
  exit 1;
}


my $MAX_RECORDS = 10;
my $scores = {};

my $excel = Spreadsheet::XLSX->new($source);
my $user_data = {};
my $users_who_rated = 0;
foreach my $sheet (@{ $excel->{Worksheet} }) {
  $sheet->{MaxRow} ||= $sheet->{MinRow};
  my $username = $sheet->{Name};
  if ($username !~ /^_/) {
    my $user = {
      raw => $sheet,
      by_slug => {}
    };
    $user_data->{$username} = $user;
    $users_who_rated++;
    foreach my $row ($skip..$MAX_RECORDS + $skip- 1) {
      my $rank = $sheet->{Cells}[$row][0]->{Val};
      my $artist = $sheet->{Cells}[$row][1]->{Val};
      my $album = $sheet->{Cells}[$row][2]->{Val};
      
      if ($rank && $artist && $album) {
        my $slug = makeSlug($artist, $album);
        my $score = ($MAX_RECORDS - $rank) + 1;
        $user->{by_slug}->{$slug} = $score;
        if (!$scores->{$slug}) {
          $scores->{$slug} = {
            points => 0,
            sources => [],
            artist => $artist,
            album => $album
          };
        }
        $scores->{$slug}->{points} += $score;
        push @{ $scores->{$slug}->{sources} }, $score;
      }
    }
  }
}

my $total_votes_cast = 0;
my $total_ratings = 0;

foreach my $slug (keys %{ $scores }) {
  # count votes for release
  $scores->{$slug}->{votes_for_item} = scalar @{ $scores->{$slug}->{sources} };
  # sum votes for release
  $scores->{$slug}->{total_score_for_item} = sum(@{ $scores->{$slug}->{sources} });
  # running total of votes
  $total_votes_cast += $scores->{$slug}->{votes_for_item};
  # running total of sums
  $total_ratings += $scores->{$slug}->{total_score_for_item};
  # average (mean) rating for release
  $scores->{$slug}->{average_rating} = $scores->{$slug}->{total_score_for_item} / $scores->{$slug}->{votes_for_item};
}

# sum average rating across releases and divide by total numer of distinct releases voted on
my $total_average_rating = sum(map { $scores->{$_}->{average_rating} } keys %{ $scores }) / scalar keys %{ $scores };
my $average_number_votes_total = $total_votes_cast / scalar keys %{ $scores };

foreach my $slug (keys %{ $scores }) {
  my $item = $scores->{$slug};
  $scores->{$slug}->{bayesian_weighted_rank} = 
    ( ($average_number_votes_total * $total_average_rating) + ($item->{votes_for_item} * $item->{total_score_for_item}) ) /
    ($average_number_votes_total + $item->{votes_for_item});
}

my $count = 0;

#my $sort_by = 'points';
my $sort_by = 'bayesian_weighted_rank';

my $usernames = join(", ", map { '@' . $_ } keys %{$user_data});

say <<"__STUFF";
  *Total Votes Cast:* $total_votes_cast
  *Total Ratings:* $total_ratings
  *Avg Rating Total:* $total_average_rating
  *Avg Votes Total:* $average_number_votes_total
  *Users Who Voted:* $users_who_rated
  *Voting Users:* $usernames
__STUFF

# say "Sorting by '$sort_by'";

foreach my $slug (sort { 
  $scores->{$b}->{$sort_by} <=> $scores->{$a}->{$sort_by}
} keys %{$scores}) {
  my $rel = $scores->{$slug};
  $rel->{average} = $rel->{points} / scalar @{ $rel->{sources} };
  $count++;
  say sprintf '%2s. *%s* - _%s_ %.01f/10 (%s votes; %.01f)', $count, $rel->{artist}, $rel->{album}, $rel->{average_rating}, $rel->{votes_for_item}, $rel->{$sort_by};
}

File: slack/node/src/process-picks.ts
import yargs from 'yargs';
import fetch from 'node-fetch';
import XLSX from 'xlsx';
import unidecode from 'unidecode';
import { sprintf } from 'sprintf-js';

const argv = yargs.argv;

interface CustomSheet {
  sheet: XLSX.WorkSheet;
  name: string;
}

const COLS = {
  RANK: 'A',
  ARTIST: 'B',
  ALBUM: 'C',
  URL: 'D',
};

function getCell(sheet, col, row): string | number | boolean {
  const cell = sheet[`${col}${row}`];
  if (!cell) {
    return false;
  }
  return cell.v;
}

function makeSlug(thing: string): string {
  return unidecode(thing)
    .toLowerCase()
    .trim()
    .replace(/[^a-z]+/, '-');
}

async function main(): Promise<string> {
  const report: string[] = [];
  const defaultURL =
    'https://docs.google.com/spreadsheets/d/e/2PACX-1vS0RBycTbJhVVtEQv4WlC1UcZpHZSyP7ym71eTiOH45NXX_3gtfFp-IQggBr0fseqavRq-thurRdvuO/pub?output=xlsx';
  const url: string = !argv.source
    ? defaultURL
    : ((argv.source as string) || defaultURL).toString();

  const skip: number =
    typeof argv.skip === 'undefined' ? 1 : parseInt(argv.skip as string, 10);

  const res = await fetch(url);
  const raw = await res.buffer();
  const wb: XLSX.WorkBook = XLSX.read(raw, { type: 'buffer' });
  const sheets = wb.SheetNames.map((sheetName: string) => {
    const sheet: XLSX.WorkSheet = wb.Sheets[sheetName];
    const cs: CustomSheet = {
      sheet,
      name: sheetName,
    };
    return cs;
  });

  const membersByName = {};
  const scores = {};

  const MAX_ALBUMS = 10;
  let usersWhoRated = 0;
  const usernames: string[] = [];
  sheets.forEach(s => {
    interface MemberRatings {
      raw: XLSX.WorkSheet;
      name: string;
      bySlug: any;
    }

    const member: MemberRatings = {
      raw: s.sheet,
      name: s.name,
      bySlug: {},
    };

    membersByName[s.name] = member;
    if (/^[^_]/.test(s.name)) {
      usersWhoRated++;
      usernames.push(s.name);
      for (let row = 1 + skip; row < MAX_ALBUMS + 1 + skip; row++) {
        const rank: number = parseInt(
          getCell(s.sheet, COLS.RANK, row).toString(),
          10,
        );
        const artist = getCell(s.sheet, COLS.ARTIST, row);
        const album = getCell(s.sheet, COLS.ALBUM, row);
        if (rank && artist && album) {
          const slug = makeSlug(`${artist}-${album}`);

          const score = MAX_ALBUMS - rank + 1;
          member.bySlug[slug] = score;
          if (!scores[slug]) {
            scores[slug] = {
              points: 0,
              sources: [],
              artist,
              album,
            };
          }
          scores[slug].points += score;
          scores[slug].sources.push(score);
        }
      }
    }
  });

  let totalVotesCast = 0;
  let totalRatings = 0;

  // foreach my $slug (keys %{ $scores }) {
  Object.keys(scores).forEach(slug => {
    scores[slug].votesForItem = scores[slug].sources.length;
    scores[slug].totalScoreForItem = scores[slug].sources.reduce(
      (p, c) => p + c,
      0,
    );
    totalVotesCast += scores[slug].votesForItem;
    totalRatings += scores[slug].totalScoreForItem;
    scores[slug].averageRating =
      scores[slug].totalScoreForItem / scores[slug].votesForItem;
  });

  const totalAverageRating =
    Object.keys(scores)
      .map(s => scores[s].averageRating)
      .reduce((p, c) => p + c, 0) / Object.keys(scores).length;
  const averageNumberVotesTotal = totalVotesCast / Object.keys(scores).length;

  // foreach my $slug (keys %{ $scores }) {
  Object.keys(scores).forEach(slug => {
    const item = scores[slug];
    scores[slug].bayesianWeightedRank =
      (averageNumberVotesTotal * totalAverageRating +
        item.votesForItem * item.totalScoreForItem) /
      (averageNumberVotesTotal + item.votesForItem);
  });

  const sortBy = 'bayesianWeightedRank';
  const allScores: any[] = [];
  Object.keys(scores).forEach(slug => {
    allScores.push(scores[slug]);
  });
  allScores.sort((a, b) => {
    if (a[sortBy] < b[sortBy]) {
      return 1;
    } else if (a[sortBy] > b[sortBy]) {
      return -1;
    }
    return 0;
  });

  report.push(`
*Total Votes Cast:* ${totalVotesCast}
*Total Ratings:* ${totalRatings}
*Avg Rating Total:* ${totalAverageRating}
*Avg Votes Total:* ${averageNumberVotesTotal}
*Users Who Voted:* ${usersWhoRated}
*Voting Users:* ${usernames
    .sort()
    .map(u => `@${u}`)
    .join(', ')}
  `);

  let count = 0;
  allScores.forEach(rel => {
    rel.averge = rel.points / rel.sources.length;
    count++;
    report.push(
      sprintf(
        '%2s. *%s* - _%s_ %.01f/10 (%s votes; %.01f)',
        count,
        rel.artist,
        rel.album,
        rel.averageRating,
        rel.votesForItem,
        rel[sortBy],
      ),
    );
  });

  console.log(report.join('\n'));
  return report.join('\n');
}

main();
